### PR TITLE
User defined batching

### PR DESCRIPTION
**Pull request by damian0815**
Add a `batch_id.txt` file containing a batch id string to a subfolder (or add a `batch_id` key to `local.yaml`) to cause images sharing the same batch id to be processed together.

### CODE REVIEW

**Comment by damian0815** [comment]
Goldkoron's tests seem to indicate that this works so i'm taking it out of draft state. Suggestions very welcome re: extra comments or refactoring to make the logic clearer.

File: data/data_loader.py [hunk 1]
```diff
@@ -125,13 +132,29 @@ def get_shuffled_image_buckets(self, dropout_fraction: float = 1.0) -> list[Imag
...
-        for bucket in buckets:
-            items.extend(buckets[bucket])
+        # handle batch_id
+        # unlabelled data (no batch_id) is in batches labelled DEFAULT_BATCH_ID.
+        items_by_batch_id = collapse_buckets_by_batch_id(buckets)
+        items = flatten_buckets_preserving_named_batch_adjacency(items_by_batch_id,
+                                                                   batch_size=batch_size,
+                                                                   grad_accum=grad_accum)
+
+        # chunk by effective batch size
```
Comment (damian0815) [inline data/data_loader.py:142]: lines 142-153 resolve the issue of aspect ratios appearing together

Comment (damian0815) [reply data/data_loader.py]: ok nm i Clean Code'd it further

**Review by victorchall — APPROVED** [review]

**Comment by victorchall** [comment]
Ok got aroud to some testing, seems not to have any perf issues, ran 2500 set with and without a few batch_id.txt

**Pull request merged**